### ISSUE TITLE
Meson does not scan generated Fortran code for module deps

### ISSUE THREAD

**Issue by adamryczkowski** [mesonbuild/meson#1348]
This is the minimum `meson.build` to reproduce the error:

```
project('simple fortran', 'fortran')
pp = find_program('python/pp_ser.py')
pp_gen = generator(pp, 
		arguments: ['-d @BUILD_DIR@', '@INPUT@'], output:'@BASENAME@.f90')
sources = ['src1.f90'   ]
generated_src = []
foreach s : sources
   generated_src += pp_gen.process(s)
endforeach
e = executable('dwarf', generated_src)
```
You also need to `touch src1.f90` and `echo "#!/usr/bin/env python">pp_ser.py` in the same directory as `meson.build`.

After `mkdir build && cd build && meson ..` I get

```
The Meson build system
Version: 0.38.0
Source dir: /home/Adama-docs/Adam/MyDocs/praca/IMGW/dev/meson_bug
Build dir: /home/Adama-docs/Adam/MyDocs/praca/IMGW/dev/meson_bug/build
Build type: native build
Project name: simple fortran
Native fortran compiler: gfortran (gcc 5.4.1)
Build machine cpu family: x86_64
Build machine cpu: x86_64
Program pp_ser.py found: YES (/usr/bin/env python /home/adam/meson_bug/pp_ser.py)
Build targets in project: 1
Traceback (most recent call last):
  File "/home/adam/.local/lib/python3.5/site-packages/mesonbuild/mesonmain.py", line 286, in run
    app.generate()
  File "/home/adam/.local/lib/python3.5/site-packages/mesonbuild/mesonmain.py", line 170, in generate
    g.generate(intr)
  File "/home/adam/.local/lib/python3.5/site-packages/mesonbuild/backend/ninjabackend.py", line 191, in generate
    self.generate_target(t, outfile)
  File "/home/adam/.local/lib/python3.5/site-packages/mesonbuild/backend/ninjabackend.py", line 386, in generate_target
    header_deps=header_deps)
  File "/home/adam/.local/lib/python3.5/site-packages/mesonbuild/backend/ninjabackend.py", line 1916, in generate_single_compile
    extra_deps += self.get_fortran_deps(compiler, abs_src, target)
  File "/home/adam/.local/lib/python3.5/site-packages/mesonbuild/backend/ninjabackend.py", line 1632, in get_fortran_deps
    with open(src) as f:
FileNotFoundError: [Errno 2] No such file or directory: '/home/adam/meson_bug/build/dwarf@exe/src1.f90'
```

I use Ubuntu 16.04 bit and meson 0.38. `python3 --version` yields `Python 3.5.2`.

**Comment by adamryczkowski** [comment]
Maybe I did something wrong (this is my first try of the meson), but at least the error message should be better ;-)

**Comment by QuLogic** [comment]
> echo "#!/usr/bin/env python">pp_ser.py

If this really is your generator, then it's somewhat obvious it failed because it doesn't create its output. Of course, the message should be a little better.

**Comment by adamryczkowski** [comment]
I think this is not an issue. The bug prevails if I upgrade the generator the following way:

```
#!/usr/bin/env python
import argparse
import os.path
parser = argparse.ArgumentParser()
parser.add_argument('-d', dest='dir')
parser.add_argument('input')
args = parser.parse_args()

out = os.path.join(args.dir, args.input)

f = open(out, 'w')
f.write('!Generated Fortran file')  # python will convert \n to os.linesep
f.close()

```

**Comment by jpakkane** [comment]
Fortran is a bit special in that it requires scanning all input files before any one of them can be compiled (to get the import dependencies). This is done at Meson configuration time. The generated files do not exist then, they only appear at compilation time. This is internal to Ninja and there does not seem to be a way to inject a scanning step between source generation and compilation. This means that generated Fortran can probably not be supported with the Ninja backend

**Comment by nirbheek** [comment]
Looks like we should only scan non-generated files then, and print a warning if any generated fortran files were found.

**Comment by adamryczkowski** [comment]
I understand. What do you think, is there anything I can do about this, or should I stop pushing for using meson at the moment?

**Comment by nirbheek** [comment]
There might be a solution to this that applies to generated sources; we're discussing it at the moment.

In the meantime, we will make it so that we don't try to scan generated files for dependencies so that you at least don't get a configuration failure.

**Comment by nirbheek** [comment]
Fwiw, if your Fortran generator only depends on the configure-time environment, you can also generate the files at configure time with [`configure_file()`](https://github.com/mesonbuild/meson/wiki/Reference-manual#configure_file). Meson will be able to scan it for `mod` dependencies then.

**Comment by adamryczkowski** [comment]
We generate virtually all Fortran input files - we have a tool that expands certain special `!$` comments into complicated calls to external library. It would help if we wouldn't need to call meson after every file edit. As I understand, meson doesn't configure in parallel, right?

**Comment by nirbheek** [comment]
Meson does not configure in parallel yet, no. If most of your Fortran files are generated, that's a problem indeed. Do many (or any) of them compile to `.mod` files? Do you use static or shared libraries in your project?

We've fixed the traceback you reported, but that still means that generated files won't be scanned for dependencies so they will be compiled in an arbitrary order. However, you can enforce a dependency order by encapsulating the output in static libraries.

As far as I know, there are no downsides to doing this if you're using at least `gfortran` or `ifort` since the linker will discard unused object files while linking. You can also force unused sections to be discarded with `-Wl,--gc-sections`.

Would this work for you?

**Comment by adamryczkowski** [comment]
At the moment all of the files compile to .mod. But we don't care about it - all we need is a bunch of executables. 

Right now I am maintaining the CMakeLists.txt that cares about the dependencies between files manually, so the thing you propose should work for us. I will get at it in a few days. In the meantime, since I am new to the meson, could you give me a hint about how to implement your advice? I would expect that the I will have a way to insert the dependencies for my executables as a tree rather than list. Or should I split the project into static libraries and treat them as the nodes of such dependency tree? Could you give me a code snippet? (I've read the manual, so you don't need to repeat that.)

**Comment by nirbheek** [comment]
Do the generated files import API from each other? If not, you don't need to do anything; just do:

```
executable('name', gensrc1, gensrc2, ...,
  ...,
  install : true)
```

Might be worth it to just try this and if you get errors we can look into the static library stuff.

**Comment by aradi** [comment]
Dear all,

has been anything done on this issue since February? I am trying to convert the make system of a middle sized scientific project to meson, and are hit by the problem that the generated (preprocessed) Fortran files are not scanned for dependencies. Or at least I think, that this is the problem since if I don't do any preprocessing, the compiling order is correct.

Most scientific Fortran projects I know use a preprocessor, so not scanning the generated source files for dependencies at all could be a show stopper for many projects. I think, there could be three possible solutions/workarounds for this problem:

* Scanning the generated source files for dependency information. If I understood this thread correctly, that can not be realized with the current ninja-framework.
* Letting the source files going through the preprocessor, when scanning for dependencies. This would make the configure phase maybe somewhat slower, but I do not think, that this is a major issue compared to the time one needs for the compilation. (Some of the preprocessors, like e.g. mine [Fypp](https://github.com/aradi/fypp) have even a Python-API, so with a tighter integration one could even spare the preprocessor startup times...)
* Alternatively, scanning the source files before the preprocessing would for most project probably still better, as not scanning them at all. Very often preprocesor commands are used for differentiating between serial and parallel code and for generating code based on templates. Very often, the dependencies (``use`` commands) are not affected by the preprocessing at all, so the dependency information could be already determined during the configuration phase without having to preprocess.

Do you think any of these approaches can be realized with Meson? It would be fantastic to have a finally a build system, which is accepted by a huge community and can be reasonably used for many scientific (Fortran) projects.

**Comment by nirbheek** [comment]
We've had some improvements to `configure_file()` in command-mode, so you should be able to use that to do your pre-processing. If, as you say, preprocessing is a fraction of the total compilation time, this turn out fine.

We don't currently keep track of mtimes for the inputs used in `configure_file()`, so we'd need to fix that so that configure is re-run when some of the inputs change, but that's easy to do. This would also allow us to cache `configure_file()` runs and make configuration faster.

**Comment by aradi** [comment]
Thanks for the quick response! if I understood it from the manual correctly, the ``configure_file()`` method only provides variable substitution in a template file, doesn't it? Or can I use it to run an arbitrary preprocessor command?

**Comment by nirbheek** [comment]
See the `command:` keyword argument. You'd use it like this:

```meson
out = configure_file(input : 'some-file',
  output : 'output-file',
  command : [pre-processor, '@INPUT@', '@OUTPUT@'])
```

**Comment by aradi** [comment]
Thanks a lot! I've tried the following:
```
project('mylib', 'fortran')

sources_fortran_basenames = ['mod1', 'mod2']

preproc_sources_fortran = []
foreach fsource_basename : sources_fortran_basenames
  infilename = '@0@.fpp'.format(fsource_basename)
  outfilename = '@0@.f90'.format(fsource_basename)
  outfile = configure_file(
      input : infilename, output : outfilename,
      command : ['cp', '@INPUT@', '@OUTPUT@'])
  preproc_sources_fortran += [outfile]
endforeach

mylib = static_library('mylib', preproc_sources_fortran)
```
But it fails, as meson apparently looks for the dependency scan in the source folder, while the preprocessed files are created by default in the build folder (where they indeed belong to):

```
The Meson build system
Version: 0.40.0
Source dir: /local/core326/aradi/volatile/mesontest
Build dir: /local/core326/aradi/volatile/mesontest/_build
Build type: native build
Project name: mylib
Native fortran compiler: gfortran (gcc 4.8.4)
Build machine cpu family: x86_64
Build machine cpu: x86_64
Configuring mod1.f90 with command
Configuring mod2.f90 with command
Build targets in project: 1
Traceback (most recent call last):
  File "/home/aradi/local/opt/virtualenv/meson/lib/python3.4/site-packages/mesonbuild/mesonmain.py", line 298, in run
    app.generate()
  File "/home/aradi/local/opt/virtualenv/meson/lib/python3.4/site-packages/mesonbuild/mesonmain.py", line 182, in generate
    g.generate(intr)
  File "/home/aradi/local/opt/virtualenv/meson/lib/python3.4/site-packages/mesonbuild/backend/ninjabackend.py", line 201, in generate
    self.generate_target(t, outfile)
  File "/home/aradi/local/opt/virtualenv/meson/lib/python3.4/site-packages/mesonbuild/backend/ninjabackend.py", line 338, in generate_target
    self.scan_fortran_module_outputs(target)
  File "/home/aradi/local/opt/virtualenv/meson/lib/python3.4/site-packages/mesonbuild/backend/ninjabackend.py", line 1710, in scan_fortran_module_outputs
    with open(filename) as f:
FileNotFoundError: [Errno 2] No such file or directory: '/local/core326/aradi/volatile/mesontest/mod2.f90'
```

Also, is there any way to use `@BASENAME@` in the `configure_file()` command? In a real project I'd prefer to store a list of file names (with extensions) instead of list of file base names.

**Comment by nirbheek** [comment]
Yes, this looks like a bug in the code that does the scanning. See [`scan_fortran_module_outputs()`](https://github.com/mesonbuild/meson/blob/1aa68cf6e3b05b63bf858aa005c34aeea362e3f9/mesonbuild/backend/ninjabackend.py#L1684) in `mesonbuild/backend/ninjabackend.py`. 

What you want to do there is to change the `filename = ` line to be:

```python
filename = s.absolute_path(self.environment.get_source_dir(),
                           self.environment.get_build_dir())
```

That should make it work for both in-source-tree files and built files. You should also change the `FIXME` above to say that it cannot work for sources generated with `custom_target()` and `generator()` because those are run after configuration is finished. I think with this it will all work. Could you try this out and make a PR if it all works for you? :)

**Comment by nirbheek** [comment]
> Also, is there any way to use @BASENAME@ in the configure_file() command? In a real project I'd prefer to store a list of file names (with extensions) instead of list of file base names.

This actually should work. See `def func_configure_file()` in `mesonbuild/interpreter.py`.

**Comment by aradi** [comment]
Sure, see PR #1788. :smiley: 

As for ``@BASENAME@``: it can only be used in the ``command`` argument of ``configure_file()``, but not in ``output``. I'll make a separate issue on that.

**Comment by scivision** [comment]
https://github.com/mesonbuild/meson/pull/1788 says it fixes this issue.
